core/pta/ocall.c

```diff
@@ -0,0 +1,344 @@
+					 &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+			if (rpc_params[n].u.memref.size > ca_param->memref.size)
+				return TEE_ERROR_BAD_PARAMETERS;
+			memcpy(ca_param->memref.buffer,
+			       PTR_ADD(mobj_va, mobj_offs),
+			       rpc_params[n].u.memref.size);
+			ca_param->memref.size = rpc_params[n].u.memref.size;
+			if (ADD_OVERFLOW(mobj_offs, rpc_params[n].u.memref.size,
+					 &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * TA Interface:
+ * -------------
+ * params[0]: VALUE IN/OUT
+ *   IN: < HOST CMD ID,  Param Types   >
+ *  OUT: < CMD Ret Val, CMD Ret Origin >
+ * params[1]: < [INOUT] TEE_Param[] >
+ * params[2]: < - >
+ * params[3]: < - >
+ *
+ * RPC Interface:
+ * --------------
+ * params[0]: VALUE IN/OUT
+ *    IN: < HOST CMD ID,      -     ,        -       >
+ *   OUT: <  Unchanged , CMD Ret Val, CMD Ret Origin >
+ * params[1]: < [IN] UUID Octets Hi, [IN] UUID Octets Lo, - >
+ * params[2..5]: < OCALL Params, if any >
+ *
+ * There are two return codes:
+ *   res: Did sending the RPC work?
+ *   ca_ret: If sending the RPC worked, what is the return value of the OCALL?
+ */
+static TEE_Result ocall_send(struct tee_ta_session *s, uint32_t param_types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t pt1 = TEE_PARAM_TYPE_GET(param_types, 0);
+	const uint32_t pt2 = TEE_PARAM_TYPE_GET(param_types, 1);
+	const uint32_t pt3 = TEE_PARAM_TYPE_GET(param_types, 2);
+	const uint32_t pt4 = TEE_PARAM_TYPE_GET(param_types, 3);
+
+	uint32_t ca_cmd_id;
+	uint32_t ca_param_types = 0;
+
+	TEE_Param *ca_params = NULL;
+	size_t ca_params_size;
+	size_t ca_params_expected_size;
+
+	struct mobj *mobj = NULL;
+	size_t mobj_sz = 0;
+	void *mobj_va = NULL;
+
+	uint32_t ca_cmd_ret;
+	uint32_t ca_cmd_ret_origin;
+
+	struct thread_param rpc_params[THREAD_RPC_MAX_NUM_PARAMS];
+	size_t rpc_num_params;
+
+	TEE_Result res;
+
+	/* Check TA interface parameter types */
+	if (pt1 != TEE_PARAM_TYPE_VALUE_INOUT ||
+	    !(pt2 == TEE_PARAM_TYPE_NONE ||
+	      pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) ||
+	    pt3 != TEE_PARAM_TYPE_NONE ||
+	    pt4 != TEE_PARAM_TYPE_NONE) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* Extract the parameters from the TA interface */
+	ca_cmd_id = params[0].value.a;
+
+	/* Process OCALL parameters, if any */
+	if (pt2 == TEE_PARAM_TYPE_MEMREF_INOUT) {
+		ca_param_types = params[0].value.b;
+		ca_params = (TEE_Param *)params[1].memref.buffer;
+		ca_params_size = params[1].memref.size;
+		ca_params_expected_size = sizeof(*ca_params) * TEE_NUM_PARAMS;
+
+		/* Check TA interface parameters */
+		if (!ca_params || ca_params_size != ca_params_expected_size)
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		/* Compute memory the CA must allocate for the OCALL params */
+		res = compute_required_mobj_size(ca_param_types, ca_params,
```

Comment:
This prevents zero copy.
How about letting the TA allocate the buffers first and pass the pointers to those in the `ca_params` above?
That would simplify this function a bit too.

Reply:
That's a good point, and I don't disagree, though I'm wondering about the API that would be exposed to TAs if we did that.

Currently, `TEE_Param` has two union'ed structures: one for a `memref` parameter and one for `value` parameter. The former currently behaves like a CA `tmpref`, while the latter behaves like a CA `value`.

What you are suggesting is instead like a CA registered `memref`, where the TA would get an SHM ID of sorts, which the OCALL PTA could then map back to its corresponding `mobj` for transport across worlds. The TA would also have to pass an offset and a size, the same way the CA has to. This would require extending `TEE_Param`.

How had you envisioned this working?

Reply:
I suppose it would be an extension which would go together with the OCALL extension.
By the way, I think all this should go into the system PTA.

Reply:
As far as I can tell, it's not trivial in that what is being suggested is registered memref-type functionality, and I don't think we have enough fields in `TEE_Param`, and associated structures, to make this work, at least not without changing them.

I'm thinking about it...

Reply:
We can take this after this PR. It's only going to affect the ABI between TA and OP-TEE, so we can let that change with regards to OCALL for a while.